File: TASK/TASK/Modules/Player/AdBreakPolicy.swift
import Foundation

/// Playback tiers understood by the ad break subsystem.
public enum SubscriptionTier: String, Equatable {
    case free
    case standard
    case premium
    case adFree
}

/// Configures how the scheduler prefetches and displays ad breaks.
///
/// The defaults here match the values shipped in the 2026.08 client rollout;
/// any change to a default is expected to go through the ads platform team.
public struct AdBreakPolicy: Equatable {
    /// How far ahead of the marker start we begin prefetching creatives.
    public let prefetchLeadTime: TimeInterval
    /// Minimum spacing between two consecutive prefetches, to avoid stampedes
    /// when several short markers sit close together.
    public let minimumPrefetchInterval: TimeInterval
    /// If the viewer scrubs backwards, only re-arm the break when the delta is
    /// larger than this threshold. Small backwards nudges (frame-accurate seeks)
    /// should not fire another prefetch.
    public let backwardSeekReArmThreshold: TimeInterval
    /// Tiers that are allowed to skip mid-roll markers even when the marker
    /// itself is not flagged skippable. Preroll is handled separately.
    public let midrollSkipAllowedTiers: Set<SubscriptionTier>

    public init(
        prefetchLeadTime: TimeInterval = 6.0,
        minimumPrefetchInterval: TimeInterval = 2.0,
        backwardSeekReArmThreshold: TimeInterval = 1.5,
        midrollSkipAllowedTiers: Set<SubscriptionTier> = [.premium, .adFree]
    ) {
        self.prefetchLeadTime = prefetchLeadTime
        self.minimumPrefetchInterval = minimumPrefetchInterval
        self.backwardSeekReArmThreshold = backwardSeekReArmThreshold
        self.midrollSkipAllowedTiers = midrollSkipAllowedTiers
    }

    public static let `default` = AdBreakPolicy()
}

File: android/player/timelinemanager/src/main/java/com/marrow/player/timelinemanager/adskip/AdBreakPolicy.kt
package com.marrow.player.timelinemanager.adskip

/**
 * Playback tiers understood by the ad break subsystem.
 */
enum class SubscriptionTier {
    FREE,
    STANDARD,
    PREMIUM,
    AD_FREE
}

/**
 * Configures how the scheduler prefetches and displays ad breaks.
 *
 * Defaults mirror the values shipped in the 2026.08 client rollout; keep the
 * iOS and Android policies in step when tuning.
 */
data class AdBreakPolicy(
    /** How far ahead of the marker start we begin prefetching creatives. */
    val prefetchLeadTimeMs: Long = 6_000L,
    /**
     * Minimum spacing between two consecutive prefetches, to avoid stampedes
     * when several short markers sit close together.
     */
    val minimumPrefetchIntervalMs: Long = 2_000L,
    /**
     * If the viewer scrubs backwards, only re-arm the break when the delta is
     * larger than this threshold. Small backwards nudges (frame-accurate seeks)
     * should not fire another prefetch.
     */
    val backwardSeekReArmThresholdMs: Long = 1_500L,
    /**
     * How long after a marker begins the viewer becomes eligible to skip it,
     * assuming their tier is entitled. The UI surfaces a "Skip in Ns" countdown
     * during this window.
     */
    val skipAvailableAfterMs: Long = 5_000L,
    /**
     * Tiers that are allowed to skip mid-roll markers even when the marker
     * itself is not flagged skippable. Preroll is handled separately.
     */
    val midrollSkipAllowedTiers: Set<SubscriptionTier> = setOf(
        SubscriptionTier.PREMIUM,
        SubscriptionTier.AD_FREE
    )
) {
    companion object {
        val DEFAULT = AdBreakPolicy()
    }
}
